### PR TITLE
Rust C API cleanup

### PR DESCRIPTION
**Pull request by andreabolognani**

### CODE REVIEW

File: src/capi/mod.rs [hunk 1]
```diff
@@ -26,33 +26,38 @@ fn virtblocks_util_build_file_name(file_name: *mut *mut c_char,
...
+}
 
 #[no_mangle]
 pub extern "C"
-fn virtblocks_devices_memballoon_set_model(memballoon: &mut devices::Memballoon,
+fn virtblocks_devices_memballoon_set_model(c_memballoon: *mut devices::Memballoon,
                                            model: devices::MemballoonModel)
 {
-    memballoon.set_model(model)
+    let rust_memballoon = unsafe { &mut *c_memballoon };
```
Comment (nertpinx) [inline src/capi/mod.rs]: My understanding here is that the unsafe here is due to the pointer de-reference, so there should be a non-null check. Even though we can just panic, I think it is safer than maybe-UB.

File: src/capi/mod.rs [hunk 2]
```diff
@@ -26,33 +26,38 @@ fn virtblocks_util_build_file_name(file_name: *mut *mut c_char,
...
+    rust_memballoon.set_model(model)
 }
 
 #[no_mangle]
 pub extern "C"
-fn virtblocks_devices_memballoon_get_model(memballoon: &devices::Memballoon) -> devices::MemballoonModel
+fn virtblocks_devices_memballoon_get_model(c_memballoon: *const devices::Memballoon) -> devices::MemballoonModel
 {
-    memballoon.get_model()
+    let rust_memballoon = unsafe { &*c_memballoon };
```
Comment (nertpinx) [inline src/capi/mod.rs]: Same here.

File: src/devices/memballoon.rs [hunk 3]
```diff
@@ -7,7 +7,6 @@ pub enum MemballoonModel {
     VirtIOTransitional,
 }
 
-#[repr(C)]
```
Comment (nertpinx) [inline src/devices/memballoon.rs:10]: Since it is not passed anywhere, that's fine.

**Review by nertpinx — COMMENTED** [review]
This needs fixes in the C API headers (`src/capi/virtblocks.h`), but that should be generated anyway, I guess, so I might look into that.

File: src/capi/mod.rs (hunk 1 above)
Comment (andreabolognani) [reply src/capi/mod.rs]: > My understanding here is that the unsafe here is due to the pointer de-reference, so there should be a non-null check. Even though we can just panic, I think it is safer than maybe-UB.

Yeah, we definitely need to do that. `assert!()`ing as done [here](http://jakegoulding.com/rust-ffi-omnibus/objects/) should work.

**Comment by andreabolognani** [comment]
> This needs fixes in the C API headers (`src/capi/virtblocks.h`)

What changes would it need? It seems to me that it's the other way around, eg. with the Rust changes now we're finally exposing the C API that the header expected to being with.

> but that should be generated anyway, I guess, so I might look into that.

I think I tried generating the header early on but couldn't get anything decent out of it... If you could make it work, or even just provide some pointers, that'd be really awesome!

**Pull request merged**